### PR TITLE
Cleanup split_scp.pl

### PR DESCRIPTION
**Pull request by xsawyerx**
Hi,

I cleaned up this script because we were trying to use it at work and it was hard to follow.

[Each commit here is self-contained, so it is easier to review it per commit than the entire diff.]

While I added a lot of whitespaces, I did make several more significant changes:

* All variables are properly declared. This prevents typos, overwriting of data, etc. that is otherwise completely undetected. (And also cleaning up memory.) Then I was able to add `strict.pm`, which is a required pragma nowadays when developing in Perl.
* Fixed insecure filehandle usage: The code made it possible to overwrite files, exhaust resources, etc. I fixed that.
* Removing bareword filehandles: This is a general bad pattern of using globally-accessible non-cleanup'able filehandles. I changed all of those to proper variables, allowing them to be scoped, cleaned up properly, and more.
* Closed filehandles explicitly. Perl attempts to close filehandles implicitly but depending on Perl version, the errors are handled differently. The safest is to close it yourself.

### CODE REVIEW

**Comment by danpovey** [comment]
that's good, but would you mind removing the whitespace just inside the ( and )?  I'd like it to look like the regular Kaldi coding style which is based on the Google style guide?  And I assume you tested this somehow?  LMK what kinds of things you did.

**Comment by xsawyerx** [comment]
I don't mind changing the whitespaces between parenthesis. That's fine. (I can try to tweak perltidy to do this.)

As for testing: There don't seem to be any testing suite in place, so the code was currently tested by adding `strict.pm` and maintaining it, meaning I didn't introduce new variables that did not exist, or removed variables that were used.

Additionally, the changes I had made, other than the security fixes and the "strict-mode" are innocuous.

Today we are testing this with our pipeline and I will update this ticket once we do, to confirm it didn't break our usage.

**Comment by xsawyerx** [comment]
After testing, we found a small thing to fix. Added a commit. I also added the whitespace changes you requested. (s/( /(/g and s/ )/)/g).

**Comment by kkm000** [comment]
@danpovey, my pet peeve with this script has been the 0-based value of the -j option, so I end up with ugly stuff like  `"scp:utils/split_scp.pl -j $nj \$((JOB-1)) lats.scp |"`. It's too late to change the semantics of the existing switch, but maybe we can add say `-k`, just like `-j`, only 1-based? `-k $nj JOB` would have been much cleaner...

**Comment by danpovey** [comment]
that makes sense, maybe we could call it -n.  need to document it very
clearly though.

**Comment by jtrmal** [comment]
Wouldn't be easier just to add a switch '--one-based'?
Y.

**Comment by jtrmal** [comment]
Oh, sorry, I realized you werent talking about making the numbers negative
but about a similar thing I was suggesting.
Y.

**Comment by kkm000** [comment]
I third `--one-based`. I thought of `k` being one 1 more than `j` was easy to remember, but `--one-based` beats it with one hand! @xsawyerx, I would really appreciate if you could add this switch.

**Comment by jtrmal** [comment]
I'm blushing :)

**Comment by xsawyerx** [comment]
I don't mind implementing it. (Honestly, I would be happy to move it to using a proper arg parser, like Getopt::Long that comes in core Perl anyway.)

However, I do think that `--one-based` is a long flag that refers to implementation detail ("in your implementation, consider "-j" to start from 1") whereas `-n` is better at saying "This is the number of workers I would like for this thing."

I apologize for jumping in late. I just saw this now.

**Comment by danpovey** [comment]
We agreed on --one-based already, so go ahead and implement that, it's not
important.

I suggest to just extend the existing code rather than use arg parsers, the
diff will be smaller and easier to check.

**Comment by xsawyerx** [comment]
I disagree this is the right way to do it, but I'll accept the group decision. I'll submit a separate PR (I think that's cleaner) that will be based on this PR. (So this one should be merged if the other is to be merged cleanly.)

**Comment by xsawyerx** [comment]
Submitted https://github.com/kaldi-asr/kaldi/pull/3162.

Meanwhile, I would like to create a branch with `Getopt::Long` to showcase how much better it would be to use a proper library to parse things.

I've tried different usages of the CLI to see if my code on that other PR worked, but found numerous inconsistencies.

```
# fails with "-j 2" at the beginning
$ perl split_scp.pl -j 2 in.scp out.scp out2.scp out3.scp
Argument "in.scp" isn't numeric in numeric lt (<) at split_scp.pl line 53.
Usage: split_scp.pl [--utt2spk=<utt2spk_file>] in.scp out1.scp out2.scp ...
 or: split_scp.pl -j num-jobs job-id [--utt2spk=<utt2spk_file>] in.scp [out.scp]
 ... where 0 <= job-id < num-jobs. at split_scp.pl line 72.

# works with "-j 2" at the end ("works" here means "ignored")
$ perl split_scp.pl in.scp out.scp out2.scp out3.scp -j 2
```

Additional weird stuff:
```
# mistake in input (not clear how it should be)
$ perl split_scp.pl -j 2 3 in.scp out.scp out2.scp out3.scp -j 2
Invalid num-jobs and job-id: 2 and 3 at split_scp.pl line 54.

# same
$ perl split_scp.pl -j 2 2 in.scp out.scp out2.scp out3.scp
Invalid num-jobs and job-id: 2 and 2 at split_scp.pl line 54.

# now it's a usage error because the parsing somehow failed
# (instead of the above examples in which we detect the user mistake)
$ perl split_scp.pl -j 2 1 in.scp out.scp out2.scp out3.scp
Usage: split_scp.pl [--utt2spk=<utt2spk_file>] in.scp out1.scp out2.scp ...
 or: split_scp.pl -j num-jobs job-id [--utt2spk=<utt2spk_file>] in.scp [out.scp]
 ... where 0 <= job-id < num-jobs. at split_scp.pl line 72.
```

**Comment by danpovey** [comment]
OK, I'll clean up this PR.  @xsawyerx you have to bear in mind that these scripts are very old.  For more recent perl script we have used `getopt`, but most scripts we write from scratch these days are in Python.  We can't change the interface of these scripts because a vast number of scripts use them.  Our main concern here is to not break anything (and to not spend too much time verifying that your changes won't break anything, hence the desire to keep your changes minimal).

**Comment by danpovey** [comment]
i mean close this PR.

**Comment by xsawyerx** [comment]
I understand your need. I did not suggest any changes to the interface whatsoever. What I'm suggesting here is to edit the implementation of the parsing to make it work correctly.

The pastes I provide above prove that if you just move arguments around, we easily confuse the parsing code. This code, in a sense, is inconsistent and can provide a less-than-optimal experience for the user.

I have quite a bit of experience with Perl and I'm willing to whip your scripts into shape. None of this involves changing their interfaces. What it does include is fixing the bugs in them and making them do what we think they should, as well as making them testable and more readable, as well as maintainable.

I'm willing to provide the results in small steps and help review all of them, to ease their integration.

None of this is to take away from your value in this: It works and it should continue to work with the same interface. I agree with it fully.
